userspace/engine/rule_loader.cpp

```diff
@@ -728,20 +662,185 @@ static void resolve_macros(
 	}
 }
 
+
+/*
+ * delim_chars
+ * helper class to look for delimiters
+ */
+struct delim_chars
+{
+	std::vector<char> m_delims = std::vector<char>(256, 0);
+
+	explicit delim_chars(const std::string &char_array)
+	{
+		for (auto c : char_array)
+		{
+			m_delims[c] = 1;
+		}
+	}
+
+	bool contains(char c) const
+	{
+		return m_delims[c]!=0;
+	}
+
+	size_t find_in(const std::string& s) const
+	{
+		for (size_t i = 0, j = s.size(); i < j; ++i)
+		{
+			if (contains(s[i]))
+			{
+				return i;
+			}
+		}
+		return std::string::npos;
+	}
+};
+
+#define LIST_DELMS "\t\n\r ,"
+
+/*
+ * list_inserter
+ */
+struct list_inserter
+{
+	using list_info_t = rule_loader::list_info;
+	using lists_map_t = indexed_vector<list_info_t>;
+
+	constexpr static const char* list_full = "\\([^\\(\\)]+\\)";
+	constexpr static const char* list_sub  = { "[^" LIST_DELMS "]+"};
+
+	const delim_chars delims{LIST_DELMS};
+
+	regex_t re_list{};
+	regex_t re_sub{};
+
+	list_inserter()
+	{
+		if (regcomp(&re_list, list_full, REG_EXTENDED) != 0)
+		{
+			ASSERT(false);
+		}
+
+		if (regcomp(&re_sub, list_sub, REG_EXTENDED) != 0)
+		{
+			ASSERT(false);
+		}
+	}
+
+	~list_inserter()
+	{
+		regfree(&re_list);
+		regfree(&re_sub);
+	}
+
+	// split string found in insert_lists by delimiters
+	// concatenate lists expansion
+	void cat_lists(std::string& ret, const std::string& cond, lists_map_t &lists) const
```

Comment:
I think this will work now, thanks. The regression tests must be able to spot any remaining issue, if any.